modules/ggl-http/src/aws_sigv4.h

```diff
@@ -0,0 +1,62 @@
+// aws-greengrass-lite - AWS IoT Greengrass runtime for constrained devices
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef GGHTTPLIB_AWS_SIGV4_H
+#define GGHTTPLIB_AWS_SIGV4_H
+
+#include "ggl/http.h"
+#include <ggl/buffer.h>
+#include <ggl/error.h>
+#include <ggl/vector.h>
+
+#define ZERO_PAYLOAD_SHA \
+    "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"
+
+#define HTTP_METHOD_GET "GET" /**< HTTP Method GET string. */
+#define HTTP_METHOD_PUT "PUT" /**< HTTP Method PUT string. */
+#define HTTP_METHOD_POST "POST" /**< HTTP Method POST string. */
+#define HTTP_METHOD_HEAD "HEAD" /**< HTTP Method HEAD string. */
+
+#define X_AMZ_SECURITY_TOKEN "x-amz-security-token"
+#define X_AMZ_DATE "x-amz-date"
+#define X_AMZ_CONTENT_SHA256 "x-amz-content-sha256"
+#define HOST_HEADER "host"
+
+// Each header key-value pair is separated by 1 ':' hence +4U.
+// Each pair of header key-value is separated from the next one by '\r\n'
+// character set - hence 2U*4U.
+#define S3_POST_EXTRA_VECT_SPACE \
+    ((sizeof(X_AMZ_CONTENT_SHA256) - 1) + (sizeof(X_AMZ_DATE) - 1) \
+     + (sizeof(X_AMZ_SECURITY_TOKEN) - 1) + (sizeof(HOST_HEADER) - 1) + 4U \
+     + (2U * 4U))
+
+
+typedef struct S3RequiredHeaders {
+    // x-amz-security-token
+    GglBuffer amz_security_token;
+    // x-amz-date
+    GglBuffer amz_date;
+    // x-amz-content-sha256
+    GglBuffer amz_content_sha256;
+    // host
+    GglBuffer host;
+} S3RequiredHeaders;
+
+/// Add HTTP headers as key:value pairs to a given vector so that it can be
+/// later on signed and added to the SigV4 header.
+GglError aws_sigv4_add_header_for_signing(
+    GglByteVec *vector, GglBuffer header_key, GglBuffer header_value
+);
+
+size_t aws_sigv4_get_iso8601_time(char buffer[17]);
```

Comment:
There is no mechanism to guarantee a C array will be at least 17 bytes passed into a function like this. Maybe move the array into a struct i.e.
```c
typedef struct Sigv4Timestamp {
    uint8_t bytes[17]; // YYYYMMDDTHHMMSSz\0
} Sigv4Timestamp;
```

Reply:
Nevermind, this isn't part of the public API

Reply:
I changed the signature of the function.